### PR TITLE
chore(demo): add `Network address` recipe page with masks for IPv6, IPv4, and MAC addresses

### PR DESCRIPTION
**Pull request by artembelik**
https://github.com/taiga-family/maskito/issues/252#issuecomment-3841274557

### CODE REVIEW

File: projects/demo/src/pages/recipes/network-address/examples/1-ipv6/mask.ts [hunk 1]
```diff
@@ -0,0 +1,8 @@
+import type {MaskitoOptions} from '@maskito/core';
+
+const hexGroup = Array.from({length: 4}, () => /[0-9A-F]/i);
+
+export default {
+    mask: Array.from({length: 8}, (_, i) => (i ? [':', ...hexGroup] : hexGroup)).flat(),
+    postprocessors: [({value, selection}) => ({value: value.toLowerCase(), selection})],
+} satisfies MaskitoOptions;
```
Comment (nsbarsukov) [inline projects/demo/src/pages/recipes/network-address/examples/1-ipv6/mask.ts:8]: Let create 
```
projects/demo-integrations/src/tests/recipes/network-address/ipv6.cy.ts
```
with the following basic test cases (feel free to extends them with your own cases):

```ts
import {DemoPath} from '@demo/constants';

describe('IPv6', () => {
    beforeEach(() => {
        cy.visit(DemoPath.NetworkAddress);
        cy.get('#ipv6 input').first().should('have.value', '').focus().as('input');
    });

    describe('valid', () => {
        it('accepts full IPv6 value and normalizes to lowercase', () => {
            cy.get('@input')
                .type('20010DB885A3000000008A2E03707334')
                .should('have.value', '2001:0db8:85a3:0000:0000:8a2e:0370:7334');
        });

        it('accepts paste of value with separators', () => {
            cy.get('@input')
                .paste('2001:0db8:0000:0000:0000:ff00:0042:8329')
                .should('have.value', '2001:0db8:0000:0000:0000:ff00:0042:8329');
        });

        it('inserts separators for partial input', () => {
            cy.get('@input').type('abcd1').should('have.value', 'abcd:1');
        });

        it('accepts mixed-case hex digits', () => {
            cy.get('@input').type('ABcdEF12').should('have.value', 'abcd:ef12');
        });
    });

    describe('invalid', () => {
        it('ignores non-hex characters', () => {
            cy.get('@input').type('abg1!2Z3').should('have.value', 'ab12:3');
        });

        it('ignores separators without enough hex digits', () => {
            cy.get('@input').type('::::').should('have.value', '');
        });

        it('does not accept more than 32 hex digits', () => {
            cy.get('@input')
                .type('f'.repeat(40))
                .should('have.value', 'ffff:ffff:ffff:ffff:ffff:ffff:ffff:ffff');
        });
    });

    describe('editing', () => {
        it('backspace deletes last character', () => {
            cy.get('@input')
                .type('2001:0db8')
                .type('{backspace}')
                .should('have.value', '2001:0db');
        });

        it('backspace also deletes trailing separator', () => {
            cy.get('@input')
                .type('2001:0')
                .type('{backspace}')
                .should('have.value', '2001');
        });

        it('delete removes next character', () => {
            cy.get('@input')
                .type('20010db8')
                .type('{leftArrow}'.repeat('db8'.length))
                .type('{del}')
                .should('have.value', '2001:0b8');
        });

        it('text selection overwrite', () => {
            cy.get('@input')
                .type('2001:0db8')
                .type('{selectall}1111')
                .should('have.value', '1111');
        });

        it('prevents insertion when full', () => {
            cy.get('@input')
                .type('1aaA2bBb3ccc4dDd5eee6fFF77778888')
                .should('have.value', '1aaa:2bbb:3ccc:4ddd:5eee:6fff:7777:8888')
                .type('a')
                .should('have.value', '1aaa:2bbb:3ccc:4ddd:5eee:6fff:7777:8888');
        });
    });
});

```

File: projects/demo/src/pages/recipes/network-address/examples/2-ipv4/mask.ts [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+export default {
+    mask: ({value}) => {
+        const octets = value.split(separator);
+
+        return octets.flatMap((octet, i) => {
+            const length = Math.max(1, Math.min(3, octet.length));
+            const group = Array.from({length}, () => /\d/);
+
+            return i < 3 ? [...group, separator] : group;
+        });
```
Comment (nsbarsukov) [inline projects/demo/src/pages/recipes/network-address/examples/2-ipv4/mask.ts]: The max number of octet groups is 4.
Should we improve mask in the following way?


```suggestion
        octets
            .map((octet, i) => {
                const length = Math.max(1, Math.min(3, octet.length));
                const group = Array.from({length}, () => /\d/);

                return i < 3 ? [...group, separator] : group;
            })
            .slice(0, 4)
            .flat();
```

File: projects/demo/src/pages/recipes/network-address/examples/2-ipv4/mask.ts [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+
+export default {
+    mask: ({value}) => {
+        const octets = value.split(separator);
+
+        return octets.flatMap((octet, i) => {
+            const length = Math.max(1, Math.min(3, octet.length));
+            const group = Array.from({length}, () => /\d/);
+
+            return i < 3 ? [...group, separator] : group;
```
Comment (nsbarsukov) [inline projects/demo/src/pages/recipes/network-address/examples/2-ipv4/mask.ts]: Now textfield always contains trailing point.
Is it really expected behavior ?

Is it better to modify mask to the following way ?
```ts
export default {
    mask: ({value}) => {
        const octets = value.split(separator).filter(Boolean);

        return octets
            .map((octet, i) => {
                const length = Math.max(1, Math.min(3, octet.length));
                const group = Array.from({length}, () => /\d/);
                const last = i === octets.length - 1;

                return last ? group : [...group, separator];
            })
            .concat([[separator, /\d/]])
            .slice(0, 4)
            .flat();
    },
    postprocessors: [
        // [...]
    ],
} satisfies MaskitoOptions;
```

What do you think about it ?

File: projects/demo/src/pages/recipes/network-address/examples/2-ipv4/mask.ts
Comment (nsbarsukov) [inline projects/demo/src/pages/recipes/network-address/examples/2-ipv4/mask.ts:1]: Could you also add some basic CY tests for `IPv4` mask too (see thread about `IPv6`) ?

Inside `projects/demo-integrations/src/tests/recipes/network-address/ipv4.cy.ts`

File: projects/demo/src/pages/recipes/network-address/examples/3-mac/mask.ts
Comment (nsbarsukov) [inline projects/demo/src/pages/recipes/network-address/examples/3-mac/mask.ts:1]: The same ask for MAC mask

```
projects/demo-integrations/src/tests/recipes/network-address/mac.cy.ts
```

File: projects/demo/src/pages/recipes/network-address/examples/2-ipv4/mask.ts
Comment (artembelik) [reply projects/demo/src/pages/recipes/network-address/examples/2-ipv4/mask.ts:1]: done

File: projects/demo/src/pages/recipes/network-address/examples/3-mac/mask.ts
Comment (artembelik) [reply projects/demo/src/pages/recipes/network-address/examples/3-mac/mask.ts:1]: done

File: projects/demo/src/pages/recipes/network-address/examples/1-ipv6/mask.ts (hunk 1 above)
Comment (artembelik) [reply projects/demo/src/pages/recipes/network-address/examples/1-ipv6/mask.ts:8]: done

File: projects/demo/src/pages/recipes/network-address/examples/2-ipv4/mask.ts (hunk 2 above)
Comment (artembelik) [reply projects/demo/src/pages/recipes/network-address/examples/2-ipv4/mask.ts]: you are right, i fixed it

File: projects/demo/src/pages/recipes/network-address/examples/2-ipv4/mask.ts (hunk 3 above)
Comment (artembelik) [reply projects/demo/src/pages/recipes/network-address/examples/2-ipv4/mask.ts]: fixed

File: projects/demo/src/pages/recipes/network-address/examples/2-ipv4/mask.ts [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+    return {elementState, data};
+};
+
+export default {
+    preprocessors: [
+        autoInsertSeparatorsPreprocessor,
+        preventSeparatorDeletionPreprocessor,
+    ],
+    mask: ({value}) => {
+        const octets = value.split(separator).filter(Boolean);
```
Comment (nsbarsukov) [inline projects/demo/src/pages/recipes/network-address/examples/2-ipv4/mask.ts]: Let's just improve logic inside dynamic mask expression instead of all new preprocessors:


```suggestion
    mask: ({value}) => {
        const octets = value
            .split(new RegExp(`(\\${separator}|\\d{${maxOctetLength}})`))
            .filter((x) => x && x !== separator);
```

Comment (nsbarsukov) [reply projects/demo/src/pages/recipes/network-address/examples/2-ipv4/mask.ts]: Could you also add tests (inside `ipv4.cy.ts`) for the such edge case (it passes with new proposed changes)

```ts
describe('partially omitted separators', () => {
    it('3 digits + 1 digit & separator + 3 digits + 3 digits', () => {
        cy.get('@input')
            .paste('1921.123123')
            .should('have.value', '192.1.123.123');
    });

    it('3 digits + 3 digits + 1 digit & separator + 3 digits', () => {
        cy.get('@input')
            .paste('1921681.001')
            .should('have.value', '192.168.1.001');
    });
});
```

File: projects/demo/src/pages/recipes/network-address/examples/2-ipv4/mask.ts [hunk 5]
```diff
@@ -0,0 +1,79 @@
+import type {MaskitoOptions, MaskitoPreprocessor} from '@maskito/core';
+
+const maxOctetValue = 255;
+const maxOctetLength = 3;
+const octetChunk = new RegExp(`.{1,${maxOctetLength}}`, 'g');
+const maxOctets = 4;
+const separator = '.';
+const digit = /\d/;
```
Comment (nsbarsukov) [inline projects/demo/src/pages/recipes/network-address/examples/2-ipv4/mask.ts]: Use SCREAMING_SNAKE_CASE for constant names

File: projects/demo/src/pages/recipes/network-address/examples/2-ipv4/mask.ts (hunk 4 above)
Comment (artembelik) [reply projects/demo/src/pages/recipes/network-address/examples/2-ipv4/mask.ts]: > Let's just improve logic inside dynamic mask expression instead of all new preprocessors:

interesting solution, corrected

> Could you also add tests (inside ipv4.cy.ts) for the such edge case (it passes with new proposed changes)

done

File: projects/demo/src/pages/recipes/network-address/examples/2-ipv4/mask.ts (hunk 5 above)
Comment (artembelik) [reply projects/demo/src/pages/recipes/network-address/examples/2-ipv4/mask.ts]: DONE :)
